// babel-plugin-transform-numeric-separator/lib/index.js
"use strict";

exports.__esModule = true;

exports.default = function () {

  function replacer(value) {
    return value.replace(/_/g, "");
  }

  function replaceNumberArg(_ref) {
    var node = _ref.node;

    if (node.callee.name === "Number") {
      node.arguments[0].value = replacer(node.arguments[0].value);
    }
  }

  var CallExpression = replaceNumberArg;
  var NewExpression = replaceNumberArg;

  return {
    inherits: _babelPluginSyntaxNumericSeparator2.default,

    visitor: {
      CallExpression: CallExpression,
      NewExpression: NewExpression,
      NumericLiteral: function NumericLiteral(_ref2) {
        var node = _ref2.node;

        if (node.extra && /_/.test(node.extra.raw)) {
          node.value = replacer(node.extra.raw);
        }
      }
    }
  };
};

var _babelPluginSyntaxNumericSeparator = require("babel-plugin-syntax-numeric-separator");

var _babelPluginSyntaxNumericSeparator2 = _interopRequireDefault(_babelPluginSyntaxNumericSeparator);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
